File: packages/node/src/launcher.ts
import { randomUUID } from "node:crypto";
import { execFile } from "node:child_process";
import { constants } from "node:fs";
import { access, mkdir, open, readFile, rm, stat } from "node:fs/promises";
import { tmpdir } from "node:os";
import { resolve } from "node:path";
import { promisify } from "node:util";

import { ConfigurationError } from "./errors.js";

const execFileAsync = promisify(execFile);

export interface LaunchPlan {
  executable: string;
  arguments: string[];
  configFile: string;
  licenseFile: string;
  driverLicenseFile?: string;
  runtimeFile?: string;
  driverRuntimeFile?: string;
  releaseRoot?: string;
  humanizeConfigFile?: string;
  nativeReadyRequestFile?: string;
  nativeReadyFile?: string;
  nativeReadyNonce?: string;
  waitForNativeReady(timeout?: number): Promise<void>;
  cleanup(): Promise<void>;
}

export interface NetworkAlignmentDiagnostic {
  configured: boolean;
  aligned: boolean;
  source?: string;
  exitIp?: string;
  missing: string[];
}

const FIRST_RELEASE_UNSUPPORTED_FEATURE_CODE = "launch_feature_unsupported";

function hasOwn(value: Record<string, unknown>, key: string): boolean {
  return Object.prototype.hasOwnProperty.call(value, key);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function rejectUnsupportedLaunchFeature(field: string): never {
  throw new ConfigurationError(
    `${field} is not supported in the first SlyBrowser release`,
    FIRST_RELEASE_UNSUPPORTED_FEATURE_CODE,
  );
}

function isUnsupportedWebRtcMode(value: unknown): boolean {
  return value === "proxy" || (isRecord(value) && value.mode === "replace");
}

function assertFingerprintProfile(options: Record<string, unknown>): void {
  const profile = isRecord(options.profile) ? options.profile : options;
  const hasMode = hasOwn(profile, "fingerprintMode");
  const hasSeed = hasOwn(profile, "fingerprintSeed");
  const hasSchema = hasOwn(profile, "fingerprintSchemaVersion");
  if (!hasMode && !hasSeed && !hasSchema) return;
  const mode = hasMode ? profile.fingerprintMode : (hasSeed || hasSchema ? "seeded" : "explicit");
  if (mode !== "explicit" && mode !== "seeded") throw new ConfigurationError("profile.fingerprintMode is invalid", "profile_invalid");
  if (hasSeed && (typeof profile.fingerprintSeed !== "string" || profile.fingerprintSeed.length === 0 || profile.fingerprintSeed.length > 128)) {
    throw new ConfigurationError("profile.fingerprintSeed is invalid", "profile_invalid");
  }
  if (hasSchema && profile.fingerprintSchemaVersion !== 1) throw new ConfigurationError("profile.fingerprintSchemaVersion is unsupported", "profile_invalid");
  if (mode === "explicit" && (hasSeed || hasSchema)) throw new ConfigurationError("profile.fingerprintSeed requires seeded fingerprint mode", "profile_invalid");
  if (mode === "seeded" && (!hasSeed || !hasSchema)) throw new ConfigurationError("Seeded fingerprint mode requires seed and schema version", "profile_invalid");
}

function assertFirstReleaseNetworkAndGeoBoundary(options: Record<string, unknown>): void {
  if (hasOwn(options, "proxy")) rejectUnsupportedLaunchFeature("proxy");
  if (hasOwn(options, "proxyAlignment")) rejectUnsupportedLaunchFeature("proxyAlignment");
  if (hasOwn(options, "geolocation")) rejectUnsupportedLaunchFeature("geolocation");
  if (hasOwn(options, "geo")) rejectUnsupportedLaunchFeature("geo");
  if (hasOwn(options, "geoIp") || hasOwn(options, "geoip")) rejectUnsupportedLaunchFeature("geoIp");
  if (hasOwn(options, "webrtc") && isUnsupportedWebRtcMode(options.webrtc)) {
    rejectUnsupportedLaunchFeature("webrtc proxy mode");
  }

  if (isRecord(options.profile)) {
    if (hasOwn(options.profile, "geolocation")) rejectUnsupportedLaunchFeature("profile.geolocation");
    if (hasOwn(options.profile, "geo")) rejectUnsupportedLaunchFeature("profile.geo");
    if (hasOwn(options.profile, "geoIp") || hasOwn(options.profile, "geoip")) rejectUnsupportedLaunchFeature("profile.geoIp");
    if (hasOwn(options.profile, "webrtc") && isUnsupportedWebRtcMode(options.profile.webrtc)) {
      rejectUnsupportedLaunchFeature("profile.webrtc proxy mode");
    }
  }
}

export function diagnoseNetworkAlignment(options: Record<string, unknown>): NetworkAlignmentDiagnostic {
  if (options.proxy !== undefined || options.proxyAlignment !== undefined) {
    return {
      configured: true,
      aligned: false,
      missing: ["proxy is not supported in the first SlyBrowser release"],
    };
  }
  return { configured: false, aligned: true, missing: [] };
}

export function normalizeNetworkSafety(options: Record<string, unknown>): Record<string, unknown> {
  const normalized = { ...options };
  assertFirstReleaseNetworkAndGeoBoundary(normalized);
  assertFingerprintProfile(normalized);
  return normalized;
}

async function privateFile(directory: string, prefix: string, payload: Buffer): Promise<string> {
  const path = resolve(directory, `${prefix}${randomUUID()}.json`);
  const handle = await open(path, constants.O_CREAT | constants.O_EXCL | constants.O_WRONLY, 0o600);
  try {
    await handle.writeFile(payload);
    await handle.sync();
  } catch (error) {
    await handle.close();
    await rm(path, { force: true });
    throw error;
  }
  await handle.close();
  await protectWindowsHandoffFile(path);
  return path;
}

async function protectWindowsHandoffFile(path: string): Promise<void> {
  if (process.platform !== "win32") return;
  const { stdout } = await execFileAsync("whoami", [], { windowsHide: true });
  const identity = stdout.trim();
  if (!identity) {
    throw new ConfigurationError("Unable to determine current Windows identity", "handoff_acl_failed");
  }
  try {
    await execFileAsync("icacls", [path, "/inheritance:r", "/grant:r", `${identity}:(F)`], { windowsHide: true });
  } catch (error) {
    await rm(path, { force: true });
    throw new ConfigurationError("Unable to restrict private handoff ACL", "handoff_acl_failed");
  }
}

function isForbiddenSecretArgument(argument: string): boolean {
  const equalsIndex = argument.indexOf("=");
  if (equalsIndex < 0) return false;
  const switchName = argument.slice(0, equalsIndex).toLowerCase();
  return switchName.includes("license") || switchName.includes("runtime");
}

function delay(milliseconds: number): Promise<void> {
  return new Promise((accept) => setTimeout(accept, milliseconds));
}

async function nativeReadyObserved(path: string, nonce: string): Promise<boolean> {
  let text: string;
  try {
    text = await readFile(path, "utf8");
  } catch (error) {
    const code = (error as NodeJS.ErrnoException).code;
    if (code === "ENOENT") return false;
    throw error;
  }
  if (!text.trim()) return false;
  let value: unknown;
  try {
    value = JSON.parse(text);
  } catch {
    return false;
  }
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new ConfigurationError("Native-ready marker is invalid", "native_ready_invalid");
  }
  const marker = value as Record<string, unknown>;
  if (marker.schemaVersion !== 1 ||
      marker.kind !== "slybrowser.native-ready" ||
      marker.ready !== true ||
      marker.nonce !== nonce) {
    throw new ConfigurationError("Native-ready marker is invalid", "native_ready_invalid");
  }
  return true;
}

async function waitForNativeReadyFile(
  path: string | undefined,
  nonce: string | undefined,
  timeout = 15_000,
): Promise<void> {
  if (path === undefined || nonce === undefined) return;
  if (!Number.isFinite(timeout) || timeout < 1) {
    throw new ConfigurationError("nativeReadyTimeout must be a positive number of milliseconds", "config_invalid");
  }
  const deadline = Date.now() + timeout;
  do {
    if (await nativeReadyObserved(path, nonce)) return;
    await delay(Math.min(50, Math.max(1, deadline - Date.now())));
  } while (Date.now() < deadline);
  throw new ConfigurationError("SlyBrowser did not report native-ready before returning", "native_ready_timeout");
}

export async function prepareLaunch(
  executable: string,
  options: Record<string, unknown>,
  lease: string | Buffer | Record<string, unknown>,
  settings: {
    tempRoot?: string | undefined;
    extraArguments?: readonly string[] | undefined;
    includeDriverLease?: boolean | undefined;
    runtimeHandoff?: Record<string, unknown> | undefined;
    driverRuntimeHandoff?: Record<string, unknown> | undefined;
    includeDriverRuntime?: boolean | undefined;
    allowRuntimeActivationTicket?: boolean | undefined;
    releaseRoot?: string | undefined;
    humanizeControl?: Record<string, unknown> | undefined;
    nativeReady?: boolean | undefined;
  } = {},
): Promise<LaunchPlan> {
  const executablePath = resolve(executable);
  try {
    await access(executablePath, constants.R_OK);
    if (!(await stat(executablePath)).isFile()) throw new Error();
  } catch {
    throw new ConfigurationError("Browser executable does not exist", "browser_missing");
  }
  const extraArguments = [...(settings.extraArguments ?? [])];
  const releaseRoot = settings.releaseRoot === undefined ? undefined : resolve(settings.releaseRoot);
  if (extraArguments.some(isForbiddenSecretArgument)) {
    throw new ConfigurationError(
      "License and runtime material must not be passed in extra browser arguments",
      "license_argument_forbidden",
    );
  }
  if (["licenseKey", "runtimeToken", "bootstrapToken", "activationTicket", "downloadTicket"].some((name) =>
      Object.prototype.hasOwnProperty.call(options, name))) {
    throw new ConfigurationError(
      "License and runtime secrets must never be placed in the browser profile handoff",
      "profile_secret_forbidden",
    );
  }
  const forbiddenRuntimeHandoffSecrets = [
    "licenseKey",
    "runtimeToken",
    ...(settings.allowRuntimeActivationTicket === true ? [] : ["activationTicket"]),
    "downloadTicket",
  ];
  for (const runtimeHandoff of [settings.runtimeHandoff, settings.driverRuntimeHandoff]) {
    if (runtimeHandoff !== undefined &&
        forbiddenRuntimeHandoffSecrets.some((name) =>
          Object.prototype.hasOwnProperty.call(runtimeHandoff, name))) {
      throw new ConfigurationError(
        "Runtime handoff must not contain long-lived keys, runtime tokens or download tickets",
        "runtime_handoff_secret_forbidden",
      );
    }
  }
  const root = resolve(settings.tempRoot ?? tmpdir());
  await mkdir(root, { recursive: true });
  const normalizedOptions = normalizeNetworkSafety(options);
  const configBytes = Buffer.from(JSON.stringify(normalizedOptions), "utf8");
  if (configBytes.length > 1024 * 1024) {
    throw new ConfigurationError("Launch configuration is too large", "config_too_large");
  }
  const leaseBytes = Buffer.isBuffer(lease)
    ? lease
    : Buffer.from(typeof lease === "string" ? lease : JSON.stringify(lease), "utf8");
  if (!leaseBytes.length || leaseBytes.length > 64 * 1024) {
    throw new ConfigurationError("License lease is missing or too large", "license_invalid_envelope");
  }
  const configFile = await privateFile(root, "sly-config-", configBytes);
  let licenseFile: string | undefined;
  let driverLicenseFile: string | undefined;
  let runtimeFile: string | undefined;
  let driverRuntimeFile: string | undefined;
  let humanizeConfigFile: string | undefined;
  let nativeReadyRequestFile: string | undefined;
  let nativeReadyFile: string | undefined;
  let nativeReadyNonce: string | undefined;
  try {
    licenseFile = await privateFile(root, "sly-license-", leaseBytes);
    if (settings.includeDriverLease) {
      driverLicenseFile = await privateFile(root, "sly-driver-license-", leaseBytes);
    }
    if (settings.runtimeHandoff !== undefined) {
      const runtimeBytes = Buffer.from(JSON.stringify(settings.runtimeHandoff), "utf8");
      if (runtimeBytes.length === 0 || runtimeBytes.length > 64 * 1024) {
        throw new ConfigurationError("Runtime handoff file is missing or too large", "runtime_handoff_invalid");
      }
      runtimeFile = await privateFile(root, "sly-runtime-", runtimeBytes);
    }
    const driverRuntimeHandoff = settings.driverRuntimeHandoff ?? (settings.includeDriverRuntime ? settings.runtimeHandoff : undefined);
    if (driverRuntimeHandoff !== undefined) {
      const driverRuntimeBytes = Buffer.from(JSON.stringify(driverRuntimeHandoff), "utf8");
      if (driverRuntimeBytes.length === 0 || driverRuntimeBytes.length > 64 * 1024) {
        throw new ConfigurationError("Runtime handoff file is missing or too large", "runtime_handoff_invalid");
      }
      driverRuntimeFile = await privateFile(root, "sly-driver-runtime-", driverRuntimeBytes);
    }
    if (settings.humanizeControl !== undefined) {
      const humanizeBytes = Buffer.from(JSON.stringify(settings.humanizeControl), "utf8");
      if (humanizeBytes.length > 64 * 1024) {
        throw new ConfigurationError("Native Humanize control file is too large", "humanize_config_too_large");
      }
      humanizeConfigFile = await privateFile(root, "sly-humanize-", humanizeBytes);
    }
    if (settings.nativeReady === true) {
      nativeReadyNonce = randomUUID();
      nativeReadyFile = await privateFile(root, "sly-native-ready-", Buffer.alloc(0));
      nativeReadyRequestFile = await privateFile(root, "sly-native-ready-request-", Buffer.from(JSON.stringify({
        schemaVersion: 1,
        kind: "slybrowser.native-ready-request",
        readyFile: nativeReadyFile,
        nonce: nativeReadyNonce,
      }), "utf8"));
    }
  } catch (error) {
    if (nativeReadyRequestFile) await rm(nativeReadyRequestFile, { force: true });
    if (nativeReadyFile) await rm(nativeReadyFile, { force: true });
    if (humanizeConfigFile) await rm(humanizeConfigFile, { force: true });
    if (driverRuntimeFile) await rm(driverRuntimeFile, { force: true });
    if (runtimeFile) await rm(runtimeFile, { force: true });
    if (driverLicenseFile) await rm(driverLicenseFile, { force: true });
    if (licenseFile) await rm(licenseFile, { force: true });
    await rm(configFile, { force: true });
    throw error;
  }
  if (!licenseFile) throw new ConfigurationError("License handoff creation failed", "license_handoff_failed");
  let cleaned = false;
  return {
    executable: executablePath,
    arguments: [
      `--sly-config-file=${configFile}`,
      `--sly-license-file=${licenseFile}`,
      ...(releaseRoot === undefined ? [] : [`--sly-release-root=${releaseRoot}`]),
      ...(runtimeFile === undefined ? [] : [`--sly-runtime-file=${runtimeFile}`]),
      ...(humanizeConfigFile === undefined ? [] : [`--sly-humanize-config=${humanizeConfigFile}`]),
      ...(nativeReadyRequestFile === undefined ? [] : [`--sly-native-ready-request-file=${nativeReadyRequestFile}`]),
      ...extraArguments,
    ],
    configFile,
    licenseFile,
    ...(releaseRoot === undefined ? {} : { releaseRoot }),
    ...(driverLicenseFile === undefined ? {} : { driverLicenseFile }),
    ...(runtimeFile === undefined ? {} : { runtimeFile }),
    ...(driverRuntimeFile === undefined ? {} : { driverRuntimeFile }),
    ...(humanizeConfigFile === undefined ? {} : { humanizeConfigFile }),
    ...(nativeReadyRequestFile === undefined ? {} : { nativeReadyRequestFile }),
    ...(nativeReadyFile === undefined ? {} : { nativeReadyFile }),
    ...(nativeReadyNonce === undefined ? {} : { nativeReadyNonce }),
    async waitForNativeReady(timeout = 15_000) {
      await waitForNativeReadyFile(nativeReadyFile, nativeReadyNonce, timeout);
    },
    async cleanup() {
      if (cleaned) return;
      cleaned = true;
      await Promise.all([
        rm(licenseFile, { force: true }),
        rm(configFile, { force: true }),
        ...(driverLicenseFile === undefined ? [] : [rm(driverLicenseFile, { force: true })]),
        ...(runtimeFile === undefined ? [] : [rm(runtimeFile, { force: true })]),
        ...(driverRuntimeFile === undefined ? [] : [rm(driverRuntimeFile, { force: true })]),
        ...(humanizeConfigFile === undefined ? [] : [rm(humanizeConfigFile, { force: true })]),
        ...(nativeReadyRequestFile === undefined ? [] : [rm(nativeReadyRequestFile, { force: true })]),
        ...(nativeReadyFile === undefined ? [] : [rm(nativeReadyFile, { force: true })]),
      ]);
    },
  };
}

File: packages/python/src/slybrowser/launcher.py
"""Secure construction and cleanup of SlyBrowser launch handoff files."""

from __future__ import annotations

import json
import os
import subprocess
import tempfile
import time
import uuid
from contextlib import contextmanager
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterator, Mapping

from .errors import ConfigurationError

FIRST_RELEASE_UNSUPPORTED_FEATURE_CODE = "launch_feature_unsupported"


@dataclass(frozen=True, slots=True)
class LaunchPlan:
    executable: Path
    arguments: tuple[str, ...]
    config_file: Path
    license_file: Path
    driver_license_file: Path | None = None
    runtime_file: Path | None = None
    driver_runtime_file: Path | None = None
    release_root: Path | None = None
    humanize_config_file: Path | None = None
    native_ready_request_file: Path | None = None
    native_ready_file: Path | None = None
    native_ready_nonce: str | None = None


def _raise_unsupported_launch_feature(field: str) -> None:
    raise ConfigurationError(
        f"{field} is not supported in the first SlyBrowser release",
        code=FIRST_RELEASE_UNSUPPORTED_FEATURE_CODE,
    )


def _is_unsupported_webrtc_mode(value: Any) -> bool:
    return value == "proxy" or isinstance(value, Mapping) and value.get("mode") == "replace"


def _assert_fingerprint_profile(options: Mapping[str, Any]) -> None:
    profile = options.get("profile") if isinstance(options.get("profile"), Mapping) else options
    has_mode = "fingerprintMode" in profile
    has_seed = "fingerprintSeed" in profile
    has_schema = "fingerprintSchemaVersion" in profile
    if not (has_mode or has_seed or has_schema):
        return
    mode = profile.get("fingerprintMode", "seeded" if has_seed or has_schema else "explicit")
    if mode not in {"explicit", "seeded"}:
        raise ConfigurationError("profile.fingerprintMode is invalid", code="profile_invalid")
    seed = profile.get("fingerprintSeed")
    if has_seed and (not isinstance(seed, str) or not seed or len(seed) > 128):
        raise ConfigurationError("profile.fingerprintSeed is invalid", code="profile_invalid")
    if has_schema and profile.get("fingerprintSchemaVersion") != 1:
        raise ConfigurationError("profile.fingerprintSchemaVersion is unsupported", code="profile_invalid")
    if mode == "explicit" and (has_seed or has_schema):
        raise ConfigurationError("profile.fingerprintSeed requires seeded fingerprint mode", code="profile_invalid")
    if mode == "seeded" and not (has_seed and has_schema):
        raise ConfigurationError("Seeded fingerprint mode requires seed and schema version", code="profile_invalid")


def _assert_first_release_network_and_geo_boundary(options: Mapping[str, Any]) -> None:
    if "proxy" in options:
        _raise_unsupported_launch_feature("proxy")
    if "proxyAlignment" in options:
        _raise_unsupported_launch_feature("proxyAlignment")
    if "geolocation" in options:
        _raise_unsupported_launch_feature("geolocation")
    if "geo" in options:
        _raise_unsupported_launch_feature("geo")
    if "geoIp" in options or "geoip" in options:
        _raise_unsupported_launch_feature("geoIp")
    if "webrtc" in options and _is_unsupported_webrtc_mode(options["webrtc"]):
        _raise_unsupported_launch_feature("webrtc proxy mode")

    profile = options.get("profile")
    if isinstance(profile, Mapping):
        if "geolocation" in profile:
            _raise_unsupported_launch_feature("profile.geolocation")
        if "geo" in profile:
            _raise_unsupported_launch_feature("profile.geo")
        if "geoIp" in profile or "geoip" in profile:
            _raise_unsupported_launch_feature("profile.geoIp")
        if "webrtc" in profile and _is_unsupported_webrtc_mode(profile["webrtc"]):
            _raise_unsupported_launch_feature("profile.webrtc proxy mode")


def diagnose_network_alignment(options: Mapping[str, Any]) -> dict[str, Any]:
    if "proxy" in options or "proxyAlignment" in options:
        return {
            "configured": True,
            "aligned": False,
            "missing": ["proxy is not supported in the first SlyBrowser release"],
        }
    return {"configured": False, "aligned": True, "missing": []}


def _normalize_network_safety(options: Mapping[str, Any]) -> dict[str, Any]:
    normalized = dict(options)
    _assert_first_release_network_and_geo_boundary(normalized)
    _assert_fingerprint_profile(normalized)
    return normalized


def _write_private_file(directory: Path, prefix: str, payload: bytes) -> Path:
    descriptor, raw_path = tempfile.mkstemp(prefix=prefix, suffix=".json", dir=directory)
    path = Path(raw_path)
    try:
        if os.name != "nt":
            os.fchmod(descriptor, 0o600)
        with os.fdopen(descriptor, "wb", closefd=True) as handle:
            handle.write(payload)
            handle.flush()
            os.fsync(handle.fileno())
    except BaseException:
        try:
            os.close(descriptor)
        except OSError:
            pass
        path.unlink(missing_ok=True)
        raise
    _protect_windows_handoff_file(path)
    return path


def _protect_windows_handoff_file(path: Path) -> None:
    if os.name != "nt":
        return
    try:
        identity = subprocess.check_output(["whoami"], text=True, stderr=subprocess.DEVNULL).strip()
        if not identity:
            raise OSError("empty Windows identity")
        subprocess.run(
            ["icacls", str(path), "/inheritance:r", "/grant:r", f"{identity}:(F)"],
            check=True,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    except Exception as error:
        path.unlink(missing_ok=True)
        raise ConfigurationError("Unable to restrict private handoff ACL", code="handoff_acl_failed") from error


def _is_forbidden_secret_argument(argument: str) -> bool:
    if "=" not in argument:
        return False
    switch_name = argument.split("=", 1)[0].lower()
    return "license" in switch_name or "runtime" in switch_name


def wait_for_native_ready(plan: LaunchPlan, *, timeout: float = 15.0) -> None:
    if plan.native_ready_file is None or plan.native_ready_nonce is None:
        return
    if not isinstance(timeout, (int, float)) or timeout <= 0:
        raise ConfigurationError("native_ready_timeout must be positive", code="config_invalid")
    deadline = time.monotonic() + float(timeout)
    while time.monotonic() < deadline:
        try:
            text = plan.native_ready_file.read_text(encoding="utf-8")
        except FileNotFoundError:
            text = ""
        if text.strip():
            try:
                marker = json.loads(text)
            except ValueError:
                marker = None
            if isinstance(marker, Mapping):
                if (
                    marker.get("schemaVersion") == 1
                    and marker.get("kind") == "slybrowser.native-ready"
                    and marker.get("ready") is True
                    and marker.get("nonce") == plan.native_ready_nonce
                ):
                    return
                raise ConfigurationError("Native-ready marker is invalid", code="native_ready_invalid")
        time.sleep(min(0.05, max(0.001, deadline - time.monotonic())))
    raise ConfigurationError("SlyBrowser did not report native-ready before returning", code="native_ready_timeout")


@contextmanager
def prepare_launch(
    executable: str | Path,
    options: Mapping[str, Any],
    lease: str | bytes | Mapping[str, Any],
    *,
    temp_root: str | Path | None = None,
    extra_arguments: tuple[str, ...] = (),
    include_driver_lease: bool = False,
    runtime_handoff: Mapping[str, Any] | None = None,
    driver_runtime_handoff: Mapping[str, Any] | None = None,
    include_driver_runtime: bool = False,
    allow_runtime_activation_ticket: bool = False,
    release_root: str | Path | None = None,
    humanize_control: Mapping[str, Any] | None = None,
    native_ready: bool = False,
) -> Iterator[LaunchPlan]:
    """Prepare one launch and remove sensitive handoff files on exit."""

    executable_path = Path(executable).resolve()
    if not executable_path.is_file():
        raise ConfigurationError("Browser executable does not exist", code="browser_missing")
    resolved_release_root = Path(release_root).resolve() if release_root is not None else None
    if any(_is_forbidden_secret_argument(argument) for argument in extra_arguments):
        raise ConfigurationError(
            "License and runtime material must not be passed in extra browser arguments",
            code="license_argument_forbidden",
        )
    if not isinstance(options, Mapping):
        raise ConfigurationError("Launch options must be an object", code="config_invalid")
    if any(secret in options for secret in ("licenseKey", "runtimeToken", "bootstrapToken", "activationTicket", "downloadTicket")):
        raise ConfigurationError(
            "License and runtime secrets must never be placed in the browser profile handoff",
            code="profile_secret_forbidden",
        )
    forbidden_runtime_secrets = ["licenseKey", "runtimeToken", "downloadTicket"]
    if not allow_runtime_activation_ticket:
        forbidden_runtime_secrets.append("activationTicket")
    for handoff in (runtime_handoff, driver_runtime_handoff):
        if handoff is not None and any(secret in handoff for secret in forbidden_runtime_secrets):
            raise ConfigurationError(
                "Runtime handoff must not contain long-lived keys, runtime tokens or download tickets",
                code="runtime_handoff_secret_forbidden",
            )

    root = Path(temp_root).resolve() if temp_root else Path(tempfile.gettempdir())
    root.mkdir(parents=True, exist_ok=True)
    normalized_options = _normalize_network_safety(options)
    config_bytes = json.dumps(normalized_options, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    if len(config_bytes) > 1024 * 1024:
        raise ConfigurationError("Launch configuration is too large", code="config_too_large")
    if isinstance(lease, Mapping):
        lease_bytes = json.dumps(lease, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    elif isinstance(lease, str):
        lease_bytes = lease.encode("utf-8")
    else:
        lease_bytes = lease
    if not isinstance(lease_bytes, bytes) or not lease_bytes or len(lease_bytes) > 64 * 1024:
        raise ConfigurationError("License lease is missing or too large", code="license_invalid_envelope")

    config_path = _write_private_file(root, "sly-config-", config_bytes)
    driver_license_path: Path | None = None
    runtime_path: Path | None = None
    driver_runtime_path: Path | None = None
    humanize_config_path: Path | None = None
    native_ready_request_path: Path | None = None
    native_ready_path: Path | None = None
    native_ready_nonce: str | None = None
    try:
        license_path = _write_private_file(root, "sly-license-", lease_bytes)
        if include_driver_lease:
            driver_license_path = _write_private_file(root, "sly-driver-license-", lease_bytes)
        if runtime_handoff is not None:
            runtime_bytes = json.dumps(runtime_handoff, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
            if len(runtime_bytes) == 0 or len(runtime_bytes) > 64 * 1024:
                raise ConfigurationError("Runtime handoff file is missing or too large", code="runtime_handoff_invalid")
            runtime_path = _write_private_file(root, "sly-runtime-", runtime_bytes)
        selected_driver_runtime_handoff = driver_runtime_handoff if driver_runtime_handoff is not None else (
            runtime_handoff if include_driver_runtime else None
        )
        if selected_driver_runtime_handoff is not None:
            driver_runtime_bytes = json.dumps(selected_driver_runtime_handoff, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
            if len(driver_runtime_bytes) == 0 or len(driver_runtime_bytes) > 64 * 1024:
                raise ConfigurationError("Runtime handoff file is missing or too large", code="runtime_handoff_invalid")
            driver_runtime_path = _write_private_file(root, "sly-driver-runtime-", driver_runtime_bytes)
        if humanize_control is not None:
            humanize_bytes = json.dumps(humanize_control, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
            if len(humanize_bytes) > 64 * 1024:
                raise ConfigurationError("Native Humanize control file is too large", code="humanize_config_too_large")
            humanize_config_path = _write_private_file(root, "sly-humanize-", humanize_bytes)
        if native_ready:
            native_ready_nonce = str(uuid.uuid4())
            native_ready_path = _write_private_file(root, "sly-native-ready-", b"")
            native_ready_request_path = _write_private_file(
                root,
                "sly-native-ready-request-",
                json.dumps(
                    {
                        "schemaVersion": 1,
                        "kind": "slybrowser.native-ready-request",
                        "readyFile": str(native_ready_path),
                        "nonce": native_ready_nonce,
                    },
                    ensure_ascii=False,
                    separators=(",", ":"),
                ).encode("utf-8"),
            )
    except BaseException:
        if native_ready_request_path is not None:
            native_ready_request_path.unlink(missing_ok=True)
        if native_ready_path is not None:
            native_ready_path.unlink(missing_ok=True)
        if "license_path" in locals():
            license_path.unlink(missing_ok=True)
        if driver_license_path is not None:
            driver_license_path.unlink(missing_ok=True)
        if runtime_path is not None:
            runtime_path.unlink(missing_ok=True)
        if driver_runtime_path is not None:
            driver_runtime_path.unlink(missing_ok=True)
        if humanize_config_path is not None:
            humanize_config_path.unlink(missing_ok=True)
        config_path.unlink(missing_ok=True)
        raise
    arguments = (
        f"--sly-config-file={config_path}",
        f"--sly-license-file={license_path}",
        *((f"--sly-release-root={resolved_release_root}",) if resolved_release_root is not None else ()),
        *((f"--sly-runtime-file={runtime_path}",) if runtime_path is not None else ()),
        *((f"--sly-humanize-config={humanize_config_path}",) if humanize_config_path is not None else ()),
        *((f"--sly-native-ready-request-file={native_ready_request_path}",) if native_ready_request_path is not None else ()),
        *extra_arguments,
    )
    plan = LaunchPlan(
        executable_path,
        tuple(arguments),
        config_path,
        license_path,
        driver_license_path,
        runtime_path,
        driver_runtime_path,
        resolved_release_root,
        humanize_config_path,
        native_ready_request_path,
        native_ready_path,
        native_ready_nonce,
    )
    try:
        yield plan
    finally:
        if native_ready_request_path is not None:
            native_ready_request_path.unlink(missing_ok=True)
        if native_ready_path is not None:
            native_ready_path.unlink(missing_ok=True)
        license_path.unlink(missing_ok=True)
        if driver_license_path is not None:
            driver_license_path.unlink(missing_ok=True)
        if runtime_path is not None:
            runtime_path.unlink(missing_ok=True)
        if driver_runtime_path is not None:
            driver_runtime_path.unlink(missing_ok=True)
        if humanize_config_path is not None:
            humanize_config_path.unlink(missing_ok=True)
        config_path.unlink(missing_ok=True)
